### ISSUE TITLE
[BUG] `package-lock` not generating properly when using "." as a workspace.

### ISSUE THREAD

**Issue by jansiegel** [npm/cli#2233]
### Current Behavior:
I'm not sure if it's a bug or lack of feature:

Is it safe to declare `root` as a package location with npm workspaces? For example:
```
"workspaces": {
  "packages": [
    ".",
    "packages/*"
  ]
}
```

I can't find anything about doing so in the RFCs or the documentation (if I missed something, point me in the right direction), so I figured it should work fine - and it does for the most part, the dependencies seem to install alright, they get linked properly, but I'm having an issue with generating `package-lock.json`.
Namely, When I declare `"."` as a package location, `package-lock` gets created, but it looks more like a slighly-modified `package.json` than a regular lock file:
```
{
  "name": "workspaces-test",
  "version": "1.0.0",
  "lockfileVersion": 2,
  "requires": true,
  "packages": {
    "": {
      "version": "1.0.0",
      "license": "ISC",
      "devDependencies": {
        "rimraf": "^2.5.4"
      },
      "workspaces": {
        "packages": [
          ".",
          "packages/*"
        ]
      }
    }
  }
}
```

Is this a bug, or should I use the root exclusively as a top-level config location?

Currently I'm using a workaround (I made a link to `"."` as `packages/rootPackage`, and I'm using that path as a workspace location), but it probably isn't the most sustainable solution.

### Expected Behavior:
If it's supported to use `"."` as a valid package location, it should generate a valid `package-lock.json`.

### Steps To Reproduce:
1. Declare `"."` as one of the workspaces
2. Run `npm i`

### Environment:
- OS: MacOS 10.15.7
- Node: 14.13.1
- npm: 7.0.13

**Comment by ruyadorno** [comment]
Thanks @jansiegel that's a very good catch! `"."` is never supposed to be used within the `workspaces` config, I'll patch that up 😊

**Comment by ljharb** [comment]
@ruyadorno why not? I’d expect, like lerna, to be able to treat the root as a workspace too.

**Comment by ruyadorno** [comment]
really? that's interesting, what's the usage/benefit of that?

**Comment by ljharb** [comment]
I do that in https://github.com/ljharb/list-exports, for example, so that when i exec a command (like the linter or tests) in every workspace, it can lint the root too.

**Comment by ruyadorno** [comment]
right 🤔 other than just executing scripts I guess there could also be usage for folks to symlink their root folder within their `node_modules` and be able to require it using package name, alright alright I'm convinced, `"."` needs to be supported then!

thanks for the quick feedback @ljharb 🙌

**Comment by jansiegel** [comment]
@ruyadorno That's awesome! 
In my case, I'm using root as a "main" library + separate workspaces in a subdirectory for the framework-specific wrappers for that library. That's basically the use case you described - I'm symlinking the root directory in `node_modules`, so the wrappers can use it.

**Comment by ruyadorno** [comment]
I had a chat with the @npm/cli team yesterday and the conclusion is that this is not something we want to just quickly patch but rather it's a signal that [**arborist**](https://github.com/npm/arborist/) is not properly handling symlinks in some cases and that warrants a more elaborate inspection/refactor of that part of the code.

@jansiegel the good news is that this use case is supported! the not so great news is that it might take a little bit more time until we get to the proper refactor and make sure that works 😊 make sure to keep an eye on this issue so that you can know once it's patched but for now I'd suggest sticking to your workaround.

**Comment by evocateur** [comment]
For the record, I really try to discourage this pattern of conflating workspace configuration and shared dev dependencies/scripts with a published package (TL;DR: it's too error prone, in my experience). I would remove the capability from Lerna, if I could, but the horse has already left the barn, so to speak.

**Comment by ljharb** [comment]
@evocateur enzyme uses it because the root defines the docs, which need to be linted with `npm run lint`, and each package also lints itself with `npm run lint`, and i want to be able to `lerna exec npm run lint` in CI and include everything. is that the pattern you find too error-prone?

**Comment by evocateur** [comment]
@ljharb If you've already invested in that particular flavor of workspace management (that is, _no_ shared dev dependencies / scripts in root, only leaf scripts/devdeps), then you're likely avoiding most of the reasons I recommend keeping the root manifest private (because the root manifest is where the dev dependencies and scripts are consolidated / run for all leaves). The issues I've observed when a root that manages the scripts etc is also public largely fall under the category of "i'ma add this new dev dependency but i forgot `--save-dev` (or i don't know the difference) && npm publish --yolo". Addressable by strong code review and so on, sure, but that's a non-trivial burden for something that could easily be avoided.

There are tradeoffs to both approaches, of course. I suppose one way of naming them distinctively (but not exhaustively, because naming is hard) would be "strong leaf" (enzyme et al) and "strong root" (my recommendation). I myself have created "strong leaf" monorepos in the past, when circumstances dictated. I won't remove support for "strong leaf" from Lerna, even in a major, largely due to historical precedent. I _will_, however, devote more effort (such as it is, lol burnout) to documenting/encouraging/supporting the "strong root" pattern of monorepo management.

So, to be clear, I'm not lodging a request to deny this behavior, I'm advocating my belief that a "strong root" should be far more common.

**Comment by ljharb** [comment]
Gotcha, thanks. I do disagree tho; I think that the only functioning monorepo is one where each subpackage looks identical to how it would if it were its own repo :-)

**Comment by ruyadorno** [comment]
This should have been fixed in: https://github.com/npm/arborist/pull/191 and should be available from the [`npm@7.1.1`](https://github.com/npm/cli/releases/tag/v7.1.1) release.
